### PR TITLE
Add Xpath Follow Links functionality (D7)

### PR DESCRIPTION
**Pull request by barbun**
# Issue
The D7 version of the module does not feature Follow Links functionality as D8 does. This is quite a handy feature that allows pagination and other dynamic links capture.

# Proposed solution
Introduce Follow Links config and logic that works the same way as the D8 version.

### CODE REVIEW

**Comment by kepol** [comment]
Thanks! I'll take a look at this.

File: quant.module [hunk 1]
```diff
@@ -520,40 +520,39 @@ function quant_seed($url, array $context = array()) {
...
   $xpath = new \DOMXPath($document);
-  $page_urls = isset($context['page_urls']) ? $context['page_urls'] : [];
+  $xpath_urls = isset($context['xpath_urls']) ? $context['xpath_urls'] : [];
 
-  $pager_xpath = array(
-    '//a[contains(@href,"page=") and contains(text(), "next")]',
-    '//a[starts-with(@href, "/") and contains(text(), "first")]',
-  );
+  $xpath_selectors = array();
+  $links_config = variable_get('quant_xpath_selectors');
```
Comment (kepol) [inline quant.module]: For existing sites, we either need this to be defaulted or we need an update hook which I think would be better.

**Review by kepol — CHANGES_REQUESTED** [review]
I still need to test but I noticed one thing.

**Comment by kepol** [comment]
Working well for me. I enabled the frontpage view and switched it to 3 per page and I got all 3 pages in Quant. So... just need the update hook and this will be ready.

Screen Shot 2023-03-02 at 12 54 05 PM

Screen Shot 2023-03-02 at 12 56 07 PM

Screen Shot 2023-03-02 at 12 56 59 PM

Screen Shot 2023-03-02 at 12 59 28 PM

File: quant.module (hunk 1 above)
Comment (barbun) [reply quant.module]: Hi @kepol thanks for reviewing this! I have now updated variable_get function to use the default value which should resolve the issue for the existing sites.

**Comment by barbun** [comment]
Added an update hook that converts old xpath variable to a new one and set defaults for sites that were not using a variable at all.

**Comment by kepol** [comment]
Thanks! I'll take a look.

**Review by kepol — APPROVED** [review]
My only concern is that the duplication of the default value may causes issues later if this is changed since it might only be changed in one place and introduce a bug but I guess it's unlikely that this will change. I still need to test this but marking approved for the code changes. There is one extra empty line but that can be fixed later.

```
    "//a[contains(@href,\"page=\") and contains(text(), \"next\")]\r\n//a[starts-with(@href, \"/\") and contains(text(), \"first\")]"),
```

**Comment by kepol** [comment]
Tested the update hook by filling in `quant_pager_xpath` and checking that `quant_xpath_selectors` got filled in with the value and `quant_pager_xpath` was then deleted from the `variable` table.

**Review by stooit — APPROVED** [review]

**Comment by kepol** [comment]
Stu also approved so merging.

**Pull request merged**